File: csharp/binary/Binary.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

public class Binary
{
    private readonly IEnumerable<int> _bits;

    public Binary(string value)
    {
        var invalid = Regex.Match(value, @"[^0-1]").Success;
        _bits = invalid ? new int[0] : value.Select(v => v == '1' ? 1 : 0);
    }

    public int ToDecimal()
    {
        return _bits.Aggregate(0, (current, bit) => current*2 + bit);
    }
}

File: fsharp/binary/Binary.fs
namespace Binary

open System.Text.RegularExpressions

type Binary(input: string) = 
    member this.toDecimal() = Seq.fold (fun acc elem -> (acc * 2) + elem) 0 this.bits

    member private this.invalid = Regex.Match(input, @"[^0-1]").Success

    member private this.bits = if this.invalid
                               then Seq.empty
                               else Seq.map this.toBit input

    member private this.toBit(c) =
        match c with
            | '1'-> 1
            | _  -> 0
